ingest/indexer/service/writelistener/bank_write_listener.go

```diff
@@ -0,0 +1,82 @@
+package writelistener
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+
+	"github.com/osmosis-labs/osmosis/osmomath"
+	"github.com/osmosis-labs/osmosis/v25/ingest/indexer/domain"
+	indexerdomain "github.com/osmosis-labs/osmosis/v25/ingest/indexer/domain"
+)
+
+var _ storetypes.WriteListener = (*bankWriteListener)(nil)
+
+type bankWriteListener struct {
+
+	// shared context to handle graceful shutdown in case of node stop.
+	ctx context.Context
+
+	client indexerdomain.PubSubClient
+
+	coldStartManager domain.ColdStartManager
+}
+
+func NewBank(ctx context.Context, client indexerdomain.PubSubClient, coldStartManager domain.ColdStartManager) storetypes.WriteListener {
+	return &bankWriteListener{
+		ctx:    ctx,
+		client: client,
+
+		coldStartManager: coldStartManager,
+	}
+}
+
+// OnWrite implements types.WriteListener.
+func (s *bankWriteListener) OnWrite(storeKey storetypes.StoreKey, key []byte, value []byte, delete bool) error {
+	if s.coldStartManager.HasIngestedInitialData() {
+		return fmt.Errorf("cold start manager has already ingested initial data")
+	}
+
+	// If the cold start manager has ingested initial data and the key is not empty and the key is a supply key.
+	if len(key) > 0 && bytes.Equal(banktypes.SupplyKey, key[:1]) {
+		denom := key[len(banktypes.SupplyKey):]
+		if err := s.publishSupply(denom, value); err != nil {
+			return err
+		}
+	}
+
+	// If the cold start manager has ingested initial data and the key is not empty and the key is a supply key.
+	if len(key) > 0 && bytes.Equal(banktypes.SupplyOffsetKey, key[:1]) {
+		denom := key[len(banktypes.SupplyOffsetKey):]
+		if err := s.publishSupply(denom, value); err != nil {
```

Comment:
Note: we will need to create a separate function for supply offsets in a followup PR

(I did not have time to finish it in the original PR)

Reply:
Followed-up here: https://github.com/osmosis-labs/osmosis/pull/8404